rust/connlib/l4-udp-dns-server/lib.rs

```diff
@@ -0,0 +1,262 @@
+    reading_udp_v6_queries: BoxStream<'static, Result<(SocketAddr, Message<Vec<u8>>)>>,
+
+    // Futures that send responses on the UDP sockets.
+    sending_udp_v4_responses: FuturesUnordered<BoxFuture<'static, Result<()>>>,
+    sending_udp_v6_responses: FuturesUnordered<BoxFuture<'static, Result<()>>>,
+
+    waker: AtomicWaker,
+}
+
+impl Server {
+    pub fn rebind_ipv4(&mut self, socket: SocketAddrV4) -> Result<()> {
+        self.udp_v4 = None;
+        self.reading_udp_v4_queries = stream::empty().boxed();
+        self.sending_udp_v4_responses.clear();
+
+        let udp_socket = Arc::new(make_udp_socket(socket)?);
+
+        self.reading_udp_v4_queries = udp_dns_query_stream(Arc::downgrade(&udp_socket));
+        self.udp_v4 = Some(udp_socket);
+
+        self.waker.wake();
+
+        tracing::info!(%socket, "Listening for UDP DNS queries");
+
+        Ok(())
+    }
+
+    pub fn rebind_ipv6(&mut self, socket: SocketAddrV6) -> Result<()> {
+        self.udp_v6 = None;
+        self.reading_udp_v6_queries = stream::empty().boxed();
+        self.sending_udp_v6_responses.clear();
+
+        let udp_socket = Arc::new(make_udp_socket(socket)?);
+
+        self.reading_udp_v6_queries = udp_dns_query_stream(Arc::downgrade(&udp_socket));
+        self.udp_v6 = Some(udp_socket);
+
+        self.waker.wake();
+
+        tracing::info!(%socket, "Listening for UDP DNS queries");
+
+        Ok(())
+    }
+
+    pub fn send_response(&mut self, to: SocketAddr, response: Message<Vec<u8>>) -> io::Result<()> {
+        let (udp_socket, workers) = match (to, self.udp_v4.clone(), self.udp_v6.clone()) {
+            (SocketAddr::V4(_), Some(socket), _) => (socket, &mut self.sending_udp_v4_responses),
+            (SocketAddr::V6(_), _, Some(socket)) => (socket, &mut self.sending_udp_v6_responses),
+            (SocketAddr::V4(_), None, _) => return Err(io::Error::other("No UDPv4 socket")),
+            (SocketAddr::V6(_), _, None) => return Err(io::Error::other("No UDPv6 socket")),
+        };
+
+        workers.push(
+            async move {
+                udp_socket
+                    .send_to(response.as_slice(), to)
+                    .await
+                    .context("Failed to send UDP response")?;
+
+                Ok(())
+            }
+            .boxed(),
+        );
+
+        Ok(())
+    }
+
+    pub fn poll(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<Query>> {
+        loop {
+            if let Poll::Ready(Some(result)) = self.sending_udp_v4_responses.poll_next_unpin(cx) {
+                result
+                    .context("Failed to send UDPv4 DNS response")
+                    .map_err(anyhow_to_io)?;
+
+                continue;
+            }
+
+            if let Poll::Ready(Some(result)) = self.sending_udp_v4_responses.poll_next_unpin(cx) {
+                result
+                    .context("Failed to send UDPv6 DNS response")
```

Comment:
If it makes things simpler, should we just stick to IPv4 for peer to peer DNS?

Reply:
Nvmd saw your comment

Reply:
I would like to in the future but in order to avoid NAT64, we first need to enforce IPv4-only DNS on the client and for that we need stop creating 1-to-1 mappings of DNS resolver IPs per defined upstream DNS server. See #8263.